### PR TITLE
Allow adding metadata

### PR DESCRIPTION
**Pull request by grzegorz-jakubiak**
Allow `AssetClient` to accept image transformations definitions as metadata

### CODE REVIEW

File: src/utils/upload-file-browser.js [hunk 1]
```diff
@@ -4,10 +4,20 @@ const { jsonApiError } = require('./simple-fetch')
 const { isFunction } = require('./typecheck')
 
 const uploadFileBrowser = (options = {}) => {
+  const metadataHeader = 'x-amz-meta-'
+  const versions = options.versions
```
Comment (jdahlke) [inline src/utils/upload-file-browser.js]: Replace `versions` with `variations` in the whole file.

One question for the metadata-header. I think we should add an additional prefix to know, that these metadata is about image converting.

Perhaps `x-amz-meta-ic-` (image converter) or `x-amz-meta-variation-` or `x-amz-meta-image-` or what you think best.

File: src/utils/upload-file-browser.js [hunk 2]
```diff
@@ -4,10 +4,20 @@ const { jsonApiError } = require('./simple-fetch')
...
+
   return new Promise((resolve, reject) => {
     const xhr = new window.XMLHttpRequest()
     xhr.open('PUT', options.uploadUrl)
     xhr.setRequestHeader('Content-Type', options.fileMimeType || '')
+
+    if (typeof versions !== 'undefined') {
```
Comment (jdahlke) [inline src/utils/upload-file-browser.js]: Just use `isEmpty(versions)`. If we don't have this package yet, just install it.

It really comes in handy. https://www.npmjs.com/package/is-empty

**Comment by grzegorz-jakubiak** [comment]
@jdahlke done and tested with `milaCRM` please have another look when you have time

File: src/clients/asset-client.js [hunk 3]
```diff
@@ -113,12 +115,21 @@ class AssetClient {
 
   __createOrUpdate (options = {}) {
     const fileName = options.file.name
+    const variations = Object.keys(options.variations).map(key => {
+      return {
+        [key]: JSON.stringify(options.variations[key])
+      }
+    }).reduce((accumulator, object) => {
+      return Object.assign(accumulator, object)
+    }, {})
```
Comment (jdahlke) [inline src/clients/asset-client.js]: Why do you have to do something like this?
I feel that could be optimized or removed.

**Comment by grzegorz-jakubiak** [comment]
@jdahlke works like a charm now! Have another look when you've got time

File: src/utils/response-helper.js [hunk 4]
```diff
@@ -7,15 +7,15 @@ function jsonApiError (response) {
...
     errors = json.errors
   } else {
     // This handles "legacy" errors from our services that haven't changed to
     // JSONAPI-compliant errors yet.
     let errorDetails = []
-    if (json.messages) {
+    if (json && json.messages) {
       errorDetails = json.messages
-    } else if (json.message) {
+    } else if (json && json.message) {
```
Comment (jdahlke) [inline src/utils/response-helper.js]: How does an XHR error response look like? Shouldn't we try to convert it to some useful error?
Right now it just always return with some unhandled error message which I implemented only as a last resort. Doesn't an XHR response contain some useful message that we can extract?

Comment (grzegorz-jakubiak) [reply src/utils/response-helper.js]: screen shot 2018-11-15 at 13 38 30

It doesn't give many information in this case.

Comment (jdahlke) [reply src/utils/response-helper.js]: That's the `response` argument in this function? Meh 😢

**Pull request merged**